tests/csrf_tests/tests.py

```diff
@@ -115,46 +128,48 @@ def test_no_csrf_cookie(self):
         """
         self._check_bad_or_missing_cookie(None, REASON_NO_CSRF_COOKIE)
 
-    def _check_bad_or_missing_token(self, token, expected):
-        """Passing None for token includes no token."""
-        if token is None:
-            req = self._get_POST_csrf_cookie_request()
-        else:
-            req = self._get_POST_request_with_token(token=token)
+    def _check_bad_or_missing_token(
+            self, expected, post_token=None, meta_token=None, token_header=None):
+        req = self._get_POST_csrf_cookie_request(
+            post_token=post_token, meta_token=meta_token, token_header=token_header,
+        )
         mw = CsrfViewMiddleware(post_form_view)
         mw.process_request(req)
         with self.assertLogs('django.security.csrf', 'WARNING') as cm:
             resp = mw.process_view(req, post_form_view, (), {})
         self.assertEqual(403, resp.status_code)
         self.assertEqual(cm.records[0].getMessage(), 'Forbidden (%s): ' % expected)
 
-    def test_csrf_cookie_no_token(self):
+    def test_csrf_cookie_bad_or_missing_token(self):
         """
-        If a CSRF cookie is present but with no token, the middleware rejects
-        the incoming request.
-        """
-        self._check_bad_or_missing_token(None, REASON_CSRF_TOKEN_MISSING)
-
-    def test_csrf_cookie_bad_token_characters(self):
-        """
-        If a CSRF cookie is present but the token has invalid characters, the
+        If a CSRF cookie is present but the token is missing or invalid, the
         middleware rejects the incoming request.
         """
-        self._check_bad_or_missing_token(64 * '*', 'CSRF token has invalid characters.')
+        cases = [
+            (None, None, REASON_CSRF_TOKEN_MISSING),
+            (16 * 'a', None, 'CSRF token from POST has incorrect length.'),
+            (64 * '*', None, 'CSRF token from POST has invalid characters.'),
+            (64 * 'a', None, 'CSRF token from POST incorrect.'),
+            (None, 16 * 'a', 'CSRF token from HTTP_X_CSRFTOKEN has incorrect length.'),
```

Comment:
Alternatively, we could change this to read:

> CSRF token from 'X-Csrftoken' header has incorrect length.

Reply:
Maybe:
```
CSRF token from the 'X-Csrftoken' HTTP header has incorrect length.
```

Reply:
Thanks, @felixxm. I updated the PR with your suggestion.